e2e_playwright/st_pydeck_chart_select_test.py

```diff
@@ -0,0 +1,223 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.conftest import ImageCompareFunction
+from e2e_playwright.shared.app_utils import click_button
+
+
+# Firefox seems to be failing but can't reproduce locally and video produces an empty page for firefox
+@pytest.mark.skip_browser("firefox")
+def test_pydeck_chart_multiselect_interactions_and_return_values(app: Page):
+    """
+    Test single selection, multi selection, and deselection all function
+    properly and return the expected values in both session_state and as a
+    return of st.pydeck.
+    """
+    # The pydeck chart takes a while to load so check that
+    # it gets attached with an increased timeout.
+    pydeck_charts = app.get_by_test_id("stDeckGlJsonChart")
+    expect(pydeck_charts).to_have_count(1, timeout=15000)
+
+    # The map assets can take more time to load, add an extra timeout
+    # to prevent flakiness.
+    app.wait_for_timeout(10000)
```

Comment:
nit: just sanity-checking whether there is something else we can look for instead of making the test at least 10 seconds long? Some DOM element or class attribute or so that gets set as soon as the assets are loaded?

Reply:
I copied this from the existing PyDeck tests: https://github.com/streamlit/streamlit/blob/develop/e2e_playwright/st_pydeck_chart_test.py#L32-L34

One of the hard parts of working with DeckGl is that all of this gets rendered into a `canvas` element, so change detection itself is hard in that element. I agree with you, and just wanted to note that I would not utilize this style of waiting usually. I can check to see if there are any other ways of doing this, but the only one I've seen is by doing detection of pixel changes on the canvas via [getImageData()](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Pixel_manipulation_with_canvas#getting_the_pixel_data_for_a_context), which also seems unwieldy in its own sense.

Reply:
I would love it if we could guarantee more stability for the Pydeck tests without requiring all these timeouts and pixel thresholds (it feels like one of the hackiest tests we have). Unfortunately, pydeck has a couple of aspects that make it complicated:
- DeckGL requires a token that we fetch from an external URL. To improve testing, we could wait for the request to succeed or we could provide a static token as part of our tests.
- The `DeckGlJsonChart` is a bit of an outlier because the component itself is not directly added to the DOM. This is caused by the `withMapboxToken` wrapper which first tries to retrieve the token before actually calling `DeckGlJsonChart` component -> I haven't really thought about it yet, but maybe there is a better way to do this so that the fetching is integrated into the `DeckGlJsonChart` component? 
- After the token is retrieved, DeckGL needs to load map resources from an external source? Is there a deterministic request we can wait for that indicates that the app loading has completed?

Reply:
See my note in this comment about testing strategy for deck.gl: https://github.com/streamlit/streamlit/pull/9377#discussion_r1764227179